### PR TITLE
use new params repo_file and new_version for upgrade tests

### PR DESCRIPTION
**Pull request by abvgedeika**
new params allow to upgrade till any version and don't use the latest packages:
1.6.2 -> 1.6.5      : repo_file = scylla.repo.upgrade_1.6 & new_version = 1.6.5
1.6.2 -> 1.7rc1     : repo_file = scylla.repo.upgrade_1.7 & new_version = 1.7rc1
1.6.2 -> 1.7 latest : repo_file = scylla.repo.upgrade_1.7 & new_version = ''

Repo file that is used in upgrade tests( from data_dir). For example:
scylla.repo.upgrade_1.6, scylla.repo.upgrade_1.7, etc
	repo_file: 'scylla.repo.upgrade'
Major scylla version to update based on 'repo_file'
For example: 1.7.rc2, 1.7.0, etc...
By default upgrade till latest major version
	new_version: ''

### CODE REVIEW

**Comment by roydahan** [comment]
@abvgedeika, let's connect this with the versions repo work we would like to do.

**Comment by roydahan** [comment]
We need to find a way to upgrade from any minor version to any minor version.
Currently, when providing a general 1-6 or 1-7 repo for example, it takes the latest.
@abvgedeika / @amoskong any suggestions?

File: data_dir/scylla.repo.upgrade_1.6 [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[scylla]
+name=Scylla for Centos $releasever - $basearch
+baseurl=https://s3.amazonaws.com/downloads.scylladb.com/rpm/centos/scylladb-1.6/$releasever/$basearch/
+enabled=1
+gpgcheck=0
+
+[scylla-generic]
+name=Scylla for centos $releasever
+baseurl=https://s3.amazonaws.com/downloads.scylladb.com/rpm/centos/scylladb-1.6/$releasever/noarch/
```
Comment (amoskong) [inline data_dir/scylla.repo.upgrade_1.6]: We have a backend service to generate and return valid repo content, maybe we can directly use that.  @roydahan @abvgedeika

**Comment by amoskong** [comment]
The problem is that we can't insure the the old packages exist in repo, we
might clean them.

So we have to backup the files manually and upgrade or downgrade from local
rpms.

File: sdcm/nemesis.py [hunk 2]
```diff
@@ -419,6 +450,28 @@ def disrupt(self):
...
+        # indexes = [x for x in range(l)]
+        # shuffle it so we will upgrade the nodes in a
+        # random order
+        # random.shuffle(indexes)
+
+        # upgrade all the nodes in random order
+        # for i in indexes:
+        #     node = self.cluster.nodes[i]
+        #     self.upgrade_node(node)
+        self.upgrade_node(self.cluster.node_to_upgrade)
```
Comment (amoskong) [inline sdcm/nemesis.py]: What's this --- 'node_to_upgrade'? do you mean the self.target_node?

File: sdcm/nemesis.py [hunk 3]
```diff
@@ -379,19 +386,43 @@ class UpgradeNemesis(Nemesis):
...
+            node.remoter.run('sudo cp /etc/scylla/scylla.yaml /etc/scylla/scylla.yaml-backup')
+            node.remoter.run('sudo nodetool snapshot')
+            node.remoter.run('sudo chown root.root /etc/yum.repos.d/scylla.repo')
+            node.remoter.run('sudo chmod 644 /etc/yum.repos.d/scylla.repo')
+            node.remoter.run('sudo yum clean all')
+            if new_version:
+                node.remoter.run('sudo yum install scylla-{0} scylla-server-{0} scylla-jmx-{0} scylla-tools-{0}'
+                                 ' scylla-conf-{0} scylla-kernel-conf-{0} scylla-debuginfo-{0} -y'.format(new_version))
+            else:
+                node.remoter.run('sudo yum update scylla scylla-server scylla-jmx scylla-tools scylla-conf scylla-kernel-conf -y')
```
Comment (amoskong) [inline sdcm/nemesis.py]: Suggestion:

```
ver_suffix = '-{}'.format(new_version) if new_version else ''
node.remoter.run('sudo yum install scylla{0} scylla-server{0} scylla-jmx{0} scylla-tools{0}'
                 ' scylla-conf{0} scylla-kernel-conf{0} scylla-debuginfo{0} -y'.format(ver_suffix))
```

File: data_dir/scylla.yaml [hunk 4]
```diff
@@ -46,6 +46,24 @@ update_db_binary: ''
 # update_db_packages: $path_of_folder_with_packages
 update_db_packages: ''
 
+es_url: http://ec2-54-89-144-171.compute-1.amazonaws.com:9200/
```
Comment (amoskong) [inline data_dir/scylla.yaml]: Let's remove this from public repo, just leave it to blank.

File: sdcm/nemesis.py [hunk 5]
```diff
@@ -379,19 +386,43 @@ class UpgradeNemesis(Nemesis):
...
-        node.remoter.run('sudo chmod 644 /etc/yum.repos.d/scylla.repo')
-        node.remoter.run('sudo yum clean all')
-        node.remoter.run('sudo yum update scylla scylla-server scylla-jmx scylla-tools scylla-conf scylla-kernel-conf -y')
+        if upgrade_node_packages:
+            # update_scylla_packages
+            node.remoter.send_files(upgrade_node_packages, '/tmp/scylla', verbose=True)
+            # node.remoter.run('sudo yum update -y --skip-broken', connect_timeout=900)
+            node.remoter.run('sudo yum install python34-PyYAML -y')
+            # replace the packages
+            node.remoter.run('yum list installed | grep scylla')
```
Comment (amoskong) [inline sdcm/nemesis.py]: `rpm -qa scylla\*` is better, yum command will involve network and not quick as rpm

File: sdcm/nemesis.py [hunk 6]
```diff
@@ -379,19 +386,43 @@ class UpgradeNemesis(Nemesis):
...
+            node.remoter.run('sudo nodetool snapshot')
+            # update *development* packages
+            node.remoter.run('sudo rpm -UvhR --oldpackage /tmp/scylla/*development* | true')
```
Comment (amoskong) [inline sdcm/nemesis.py]: 1. Can we replace the `true` with grace (ignore_status=True)?

2. How about remove development pattern? then we can install anything.
such as some minor release that doesn't exists in repo, this provide a way to upgrade/downgrade of any minor release.   @abvgedeika @roydahan

**Comment by amoskong** [comment]
Talked to Andrei, we can merge the worked PR first, improve it by another PR. Because we might use it in testing packages.

File: sdcm/nemesis.py (hunk 2 above)
Comment (abvgedeika) [inline sdcm/nemesis.py]: no, self.node_to_upgrade is a node that was selected for upgrade https://github.com/scylladb/scylla-cluster-tests/blob/master/upgrade_test.py#L655 . I node per nemesis

**Pull request merged**